### PR TITLE
MicroSplit API: noise model & loss refactoring

### PR DESCRIPTION
**Pull request by CatEek**
## Disclaimer

- [ ] I am an AI agent.
- [x] I have used AI and I thoroughly reviewed every line.
- [ ] I have not used AI extensively.

## Description

> [!NOTE]  
> **tldr**: Refactor MicroSplit noise-model and loss configuration, replacing
> fragmented likelihood configs with a unified logic.

### Background - why do we need this PR?

The previous implementation spread noise-model related code across 3 separate
Pydantic models (`GaussianLikelihoodConfig`, `NMLikelihoodConfig`, and
`MultiChannelNMConfig`), three loss types (`musplit`, `denoisplit`,
`denoisplit_musplit`), and multiple `VAEModule` attributes
(`gaussian_likelihood`, `noise_model_likelihood`). This made it hard to reason
about which combination of configs was valid, made code very hard to read and error prone.

### Overview - what changed?

- Likelihood configuration objects are removed; their parameters live directly
  in `LVAELossConfig`.
- `predict_logvar` is now a plain `bool` everywhere (was `Literal[None,
  "pixelwise"]`).
- `GaussianMixtureNMConfig` exposes a `from_npz` classmethod instead of
  accepting a `path` field that triggered side-effectful validation.(That is questionable and likely to be changed)
- `VAEModule` holds data statistics directly and passes them into the loss
  function, removing the intermediary likelihood objects.
- `NoiseModelTrainer` is added(Not yet used).
- musplit_loss, denoisplit_loss, and denoisplit_musplit_loss are replaced
by a single microsplit_loss, controlled by `musplit_weight` / `denoisplit_weight`.. The SupportedLoss enum retains only hdn and
microsplit.

Performance tested on a subset of ht_lif24

### Please ensure your PR meets the following requirements:
- [x] Code builds and passes tests locally, including doctests
- [x] New tests have been added (for bug fixes/features)
- [ ] Pre-commit passes
- [ ] PR to the documentation exists (for bug fixes / features)

### CODE REVIEW

File: src/careamics/config/algorithms/vae_algorithm_config.py
Comment (jdeschamps) [inline src/careamics/config/algorithms/vae_algorithm_config.py:1]: Given that most of the validation in this file has to do with the parameters being either for HDN or MicroSplit, I think we should split the two configurations.

Comment (jdeschamps) [reply src/careamics/config/algorithms/vae_algorithm_config.py:1]: That can be a subsequent PR

File: src/careamics/config/algorithms/vae_algorithm_config.py [hunk 1]
```diff
@@ -136,24 +116,15 @@ def predict_logvar_validation(self: Self) -> Self:
...
-        #         self.model.predict_logvar is None
-        #     ), "Model `predict_logvar` must be `None` for algorithm `hdn`."
-        #     if self.gaussian_likelihood is not None:
-        #         assert self.gaussian_likelihood.predict_logvar is None, (
-        #             "Gaussian likelihood model `predict_logvar` must be `None` "
-        #             "for algorithm `hdn`."
-        #         )
-        # TODO check this
+        # Check consistency between model.predict_logvar and loss.predict_logvar
+        model_predicts_logvar = self.model.predict_logvar
```
Comment (jdeschamps) [inline src/careamics/config/algorithms/vae_algorithm_config.py:120]: why are there two predict_logvar? where are they used?

File: src/careamics/config/architectures/lvae_config.py [hunk 2]
```diff
@@ -39,7 +39,7 @@ class LVAEConfig(ArchitectureConfig):
         default="ELU",
     )
 
-    predict_logvar: Literal[None, "pixelwise"] = "pixelwise"
+    predict_logvar: bool = True
```
Comment (jdeschamps) [inline src/careamics/config/architectures/lvae_config.py:42]: can we add a parameter description?

File: src/careamics/config/losses/loss_config.py [hunk 3]
```diff
@@ -6,24 +6,16 @@
...
-    aggregation: Literal["sum", "mean"] = "mean"
-    """Aggregation of the KL loss across different layers."""
-    free_bits_coeff: float = 0.0
-    """Free bits coefficient for the KL loss."""
-    annealing: bool = False
-    """Whether to apply KL loss annealing."""
-    start: int = -1
-    """Epoch at which KL loss annealing starts."""
-    annealtime: int = 10
-    """Number of epochs for which KL loss annealing is applied."""
```
Comment (jdeschamps) [inline src/careamics/config/losses/loss_config.py:26]: Can you comment on why were these parameters removed? Are they never changed and were just left over from over-parametrization?

File: src/careamics/config/noise_model/noise_model_config.py
Comment (jdeschamps) [inline src/careamics/config/noise_model/noise_model_config.py:1]: I don't see the point of having a Pydantic model for a list of single channel NMs

File: src/careamics/losses/lvae/losses.py [hunk 4]
```diff
@@ -372,205 +343,137 @@ def musplit_loss(
...
-    -------
-    output : Optional[dict[str, torch.Tensor]]
-        A dictionary containing the overall loss `["loss"]`, the reconstruction loss
-        `["reconstruction_loss"]`, and the KL divergence loss `["kl_loss"]`.
-    """
-    assert noise_model_likelihood is not None
 
-    predictions, td_data = model_outputs
+    elif gaussian_weight > 0:
+        recons_loss = -_compute_gaussian_log_likelihood(
```
Comment (jdeschamps) [inline src/careamics/losses/lvae/losses.py:413]: It seems to me that `_compute_gaussian_log_likelihood` and the other `_compute_blahblah` are always negated (multiplied by `-1` rather). Why not do that directly in the function?

File: src/careamics/losses/lvae/losses.py [hunk 5]
```diff
@@ -2,31 +2,117 @@
...
+    data_std_tensor = torch.as_tensor(
+        data_std, dtype=torch.float32, device=reconstruction.device
+    )
+
+    # Move noise model to correct device if needed
+    if reconstruction.device != noise_model.device:
+        noise_model.to_device(reconstruction.device)
+
+    # Denormalize predictions and targets
+    reconstruction_denorm = reconstruction * data_std_tensor + data_mean_tensor
```
Comment (jdeschamps) [inline src/careamics/losses/lvae/losses.py:96]: I still believe the noise model itself could be normalized, let's open an issue.

File: src/careamics/losses/lvae/losses.py [hunk 6]
```diff
@@ -372,205 +343,137 @@ def musplit_loss(
...
+    if gaussian_weight > 0 and not config.predict_logvar:
+        raise ValueError(
+            "predict_logvar must be True in config when musplit_weight > 0"
+        )
+    if nm_weight > 0 and (noise_model is None or data_mean is None or data_std is None):
+        raise ValueError(
+            "noise_model, data_mean, and data_std required when denoisplit_weight > 0"
+        )
+    recons_loss: torch.Tensor | float = 0.0
+    if nm_weight > 0 and gaussian_weight > 0:
```
Comment (jdeschamps) [inline src/careamics/losses/lvae/losses.py:377]: Since there are multiple `if clause` with the same conditions, it would make sense to split them into different functions.

File: src/careamics/losses/lvae/losses.py
Comment (jdeschamps) [inline src/careamics/losses/lvae/losses.py:1]: I feel the loss could be further refactored in a subsequent PR.

File: src/careamics/models/lvae/noise_models.py [hunk 7]
```diff
@@ -93,28 +93,12 @@ def noise_model_factory(
...
-                # trained_nm = train_gm_noise_model(model_config)
-                # return trained_nm
-                raise NotImplementedError(
-                    "GaussianMixtureNoiseModel model training is not implemented."
-                )
-            else:
-                raise NotImplementedError(
-                    f"Model {model_config.model_type} is not implemented"
-                )
+        if model_config.model_type == "GaussianMixtureNoiseModel":
```
Comment (jdeschamps) [inline src/careamics/models/lvae/noise_models.py:96]: That string should indeed be simplified

File: src/careamics/models/lvae/noise_models.py [hunk 8]
```diff
@@ -268,12 +234,51 @@ def likelihood(self, obs: torch.Tensor, signal: torch.Tensor) -> torch.Tensor:
...
+        NDArray
+            Sampled noisy observation with same shape as input signal.
+        """
+        if signal.ndim < 3:
+            raise ValueError(
+                f"Signal must have at least 3 dimensions (C, Y, X), got {signal.ndim}D"
+            )
+
+        n_channels = signal.shape[-3]
+        if n_channels != self._nm_cnt:
```
Comment (jdeschamps) [inline src/careamics/models/lvae/noise_models.py:260]: `self._nm_cnt` is not an easy name to read or understand.

File: src/careamics/prediction_utils/lvae_prediction.py [hunk 9]
```diff
@@ -1,11 +1,24 @@
-"""Module containing pytorch implementations for obtaining predictions from an LVAE."""
+"""Module containing pytorch implementations for obtaining predictions from an LVAE.
 
-from typing import Any
+DEPRECATED: This module uses the old likelihood-based approach and needs to be updated
```
Comment (jdeschamps) [inline src/careamics/prediction_utils/lvae_prediction.py:3]: Does that mean that prediction is not compatible with the current state of this branch?

**Review by jdeschamps — CHANGES_REQUESTED** [review]

File: src/careamics/config/algorithms/vae_algorithm_config.py (hunk 1 above)
Comment (jdeschamps) [reply src/careamics/config/algorithms/vae_algorithm_config.py:120]: They are indeed used in both, let's think later how that can be improved.

File: src/careamics/config/losses/loss_config.py (hunk 3 above)
Comment (CatEek) [reply src/careamics/config/losses/loss_config.py:26]: Yes, that is the case

File: src/careamics/config/architectures/lvae_config.py (hunk 2 above)
Comment (CatEek) [reply src/careamics/config/architectures/lvae_config.py:42]: done

File: src/careamics/models/lvae/noise_models.py (hunk 8 above)
Comment (CatEek) [reply src/careamics/models/lvae/noise_models.py:260]: This might have to be addressed to original author

File: src/careamics/losses/lvae/losses.py (hunk 4 above)
Comment (CatEek) [reply src/careamics/losses/lvae/losses.py:413]: At this point I dont think so because that would defeat the purpose of having single microsplit loss

File: src/careamics/prediction_utils/lvae_prediction.py (hunk 9 above)
Comment (CatEek) [reply src/careamics/prediction_utils/lvae_prediction.py:3]: Not for MS, they might have some utility for HDN, so I'd leave it here for now

File: src/careamics/config/architectures/lvae_config.py (hunk 2 above)
Comment (CatEek) [reply src/careamics/config/architectures/lvae_config.py:42]: Done

File: src/careamics/config/noise_model/noise_model_config.py
Comment (CatEek) [reply src/careamics/config/noise_model/noise_model_config.py:1]: NMs need further refactoring anyway

File: src/careamics/models/lvae/noise_models.py (hunk 7 above)
Comment (CatEek) [reply src/careamics/models/lvae/noise_models.py:96]: I'd say it should go to the same bucket as MultiChannel NM, needs refactoring

**Comment by jdeschamps** [comment]
Closing https://github.com/CAREamics/careamics/pull/700

**Comment by CatEek** [comment]
All pre-commit issues are solved in #940, they all relate to noise_model_trainer.py module and one deprecated lvae_prediction.py

File: src/careamics/models/lvae/noise_models.py (hunk 8 above)
Comment (jdeschamps) [reply src/careamics/models/lvae/noise_models.py:260]: I'd rather address them to whoever is refactoring that part of the code.
